rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+- It will also not contains test plugin apps, making it really hard to FTR
+- AST parsing is a complex mechanism for an already unsatisfactory alternative
+ 
+### Duplicated server-side `application.register` API 
+ 
+One other option would be to duplicate the `application.register` API on the server side, with a subset of the 
+client-side metadata.
+
+```ts
+core.application.register({
+  id: 'app_status',
+  title: 'App Status',
+  euiIconType: 'snowflake'
+});
+```
+
+This way, the applications could be searchable from the server using this server-side `applications` registry.
+
+However
+- It forces plugin developers to add this API call. In addition to be a very poor developer experience, it can also
+  very easily be forgotten, making a given app non searchable
+- client-side only plugins would need to add a server-side part to their plugin just to register their application on
+  the server side
+  
+## `GlobalSearchResult.url` could be a struct instead of a url for internal results.
+
+The initial idea for `GlobalSearchResult.url` was to have a polymorphic structure instead of just a string:
+
+```ts
+url: { absUrl: string } | { application: string; path?: string };
+```
+
+That was making it way easier to redirect to an internal result from the UI, as we could directly call
+`application.navigateTo(application, { path })`.
+
+However, that didn't bring answer for the (future) need to be able to search for and redirect to object in 
+different spaces. We could then have changed the structure to
+
+```ts
+url: { absUrl: string } | { application: string; path?: string, space?: string };
+```
+
+But this had some caveats:
+- `space` is an xpack plugin, adding this `space` property in an oss part of the code is problematic
+- The `space` API is not callable from core or oss, meaning that we would have to 'forge' the url to this space anyway
+- this is really not generic. If another plugin was to alter the basepath in another way, we would have needed to add it another property 
+
+So even if the 'parsable absolute url' approach seems fragile, it still felt better than this alternative.
+
+## We could use plain string instead of an enum for `GlobalSearchResult.type`
+
+The current static enum used for type
+
+```ts
+enum SEARCH_TYPE {
+  // non-exhaustive
+  applications = 10,
+  dashboard = 20,
+  visualization = 30,
+  search = 40,
+}
+
+/** @public */
+type GlobalSearchResultType = keyof typeof SEARCH_TYPE;
+
+interface GlobalSearchResult {
+  // [...]
+  type: GlobalSearchResultType;
+}
+```
+
+has some limitations:
+- It forces the enum to be modified every time a new type is added
+- 3rd party plugins cannot introduce new types
+
+We could change the API to accept plain strings for `GlobalSearchResult.type`. However, atm this enum approach 
+is needed as the ordinal values of the entries is used in results sorting. Changing to plain strings forces to find 
```

Comment:
Can't we allow results providers themselves to specify the order, similar to how we allow [applications currently](https://github.com/elastic/kibana/blob/48a33abdeed147932401cc3a24c36669189f67f3/src/core/public/application/types.ts#L106)?

Reply:
++ to this and using strings so plugins can register with order specified.

Reply:
++

Reply:
@kobelb that was an option I pointed out in the initial discussion but removed here. However depending on the decision in https://github.com/elastic/kibana/pull/64284#discussion_r414581051, this maybe obsolete, as if we provide only a result-streaming API, core will no longer handle sorting.

Reply:
@kobelb 

> Can't we allow results providers themselves to specify the order

Now that I think more about it, I'm still unsure what the suggestion exactly is. Did you mean that providers should be able to return an `order` for a given `type`, for the `resultProvider` itself, or for individual results.

As multiple result providers would be able to return results of the same type (I.E the base SO provider and the 'cloud' SO provider will both return 'dashboard' types), there is not a 1-1 relation between a provider and a type, so they can't have this responsibility.

Should the `GlobalSearchResultProvider` type have an `order` field instead? In that case, as we will likely no longer handle sorting from the GS service itself (see https://github.com/elastic/kibana/pull/64284#discussion_r414581051), this property should probably be reflected on the `GlobalSearchResult` as something like `providerOrder`?

Reply:
My suggestion was initially meant to replace the integer component of the enum so plugins could register results providers without having to modify the centralized enum. 

You bring up a good point though that since we're now forcing the consumer to order the results, perhaps we don't need this at all.

Reply:
https://github.com/elastic/kibana/pull/64284/commits/24b1ec3bc4e7a1041f64f1c7a8ecc34ce60aed29 uses plain strings instead of an enum for `GlobalSearchResult.type`. Changed the proposed sort algorithm. 

Note that sorting still depends on the decision we take regarding https://github.com/elastic/kibana/pull/64284#discussion_r414581051

Reply:
Done in https://github.com/elastic/kibana/pull/64284/commits/d2289f79e44ac6d042a049ee307395b25df5913b